server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/DialectService.java

```diff
@@ -60,17 +60,48 @@ public DialectService(CopybookService copybookService,
    * Process the source file text with dialects
    *
    * @param dialects the list of enabled dialects
-   * @param context is a DialectProcessingContext class with all needed data for dialect processing
+   * @param context  is a DialectProcessingContext class with all needed data for dialect processing
    * @return dialects outcome
    */
   public ResultWithErrors<DialectOutcome> process(List<String> dialects, DialectProcessingContext context) {
-    return dialects.stream()
-        .map(this::getDialectByName)
-        .reduce(
-            ResultWithErrors.of(new DialectOutcome(context.getText(), ImmutableList.of(), ImmutableMultimap.of())),
-            (previousResult, dialect) -> processDialect(previousResult, dialect, context),
-            DialectService::mergeResults
-        );
+    List<CobolDialect> orderedDialects = sortDialects(dialects);
+    List<SyntaxError> errors = new LinkedList<>();
+    for (CobolDialect orderedDialect : orderedDialects) {
+      errors.addAll(orderedDialect.extend(context));
+      context.getExtendedSource().commitTransformations();
+    }
+    ResultWithErrors<DialectOutcome> acc = new ResultWithErrors(
+        new DialectOutcome(context), errors);
+    for (CobolDialect orderedDialect : orderedDialects) {
+      acc = processDialect(acc, orderedDialect, context);
+      context.getExtendedSource().commitTransformations();
+    }
+    return acc;
+  }
+
+  private LinkedList<CobolDialect> sortDialects(List<String> dialects) {
+    LinkedList<CobolDialect> orderedDialects = new LinkedList<>();
+    LinkedList<String> dialectsQueue = new LinkedList<>(dialects);
+    while (!dialectsQueue.isEmpty()) {
+      CobolDialect dialect = getDialectByName(dialectsQueue.pop());
+      if (dialect.runBefore().isEmpty()) {
+        orderedDialects.add(dialect);
+      } else {
+        for (String name : dialect.runBefore()) {
+          CobolDialect d = getDialectByName(name);
+          int index = orderedDialects.indexOf(d);
+          if (index >= 0) {
+            orderedDialects.add(index, dialect);
+          } else {
+            if (!dialectsQueue.contains(d.getName())) {
+              dialectsQueue.add(d.getName());
+            }
+            dialectsQueue.add(dialect.getName());
```

Comment:
Can this lead to an infinite loop, or this would never happen in our context?

What I mean is, a dialect list of [C, B, A( runs before X)]

Reply:
In fact, this method looks weird and should be moved to a separate class and tested by unit tests. Let's do it in a different PR